core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java

```diff
@@ -528,22 +533,38 @@ private TaskInfo doUpdateTask(
             }
         }
 
-        fragment.map(PlanFragment::getActiveCatalogs)
-                .ifPresent(activeCatalogs -> {
-                    Set<CatalogHandle> catalogHandles = activeCatalogs.stream()
-                            .map(CatalogProperties::getCatalogHandle)
-                            .collect(toImmutableSet());
-                    if (sqlTask.setCatalogs(catalogHandles)) {
-                        ReentrantReadWriteLock.ReadLock catalogInitLock = catalogsLock.readLock();
-                        catalogInitLock.lock();
-                        try {
-                            connectorServicesProvider.ensureCatalogsLoaded(session, activeCatalogs);
-                        }
-                        finally {
-                            catalogInitLock.unlock();
-                        }
+        Optional<List<CatalogProperties>> activeCatalogsOpt = fragment.map(PlanFragment::getActiveCatalogs);
+        if (activeCatalogsOpt.isPresent()) {
+            List<CatalogProperties> activeCatalogs = activeCatalogsOpt.get();
+            Set<CatalogHandle> catalogHandles = activeCatalogs.stream()
+                    .map(CatalogProperties::getCatalogHandle)
+                    .collect(toImmutableSet());
+            if (sqlTask.setCatalogs(catalogHandles)) {
+                ListenableFuture<Void> catalogLoading = Futures.submit(() -> {
+                    ReentrantReadWriteLock.ReadLock catalogInitLock = catalogsLock.readLock();
+                    catalogInitLock.lock();
```

Comment:
why lock is needed? Can catalogs be updated concurrently? If so, how do you guarantee that they are updated in correct order?

Reply:
> The lock ensures proper ordering

What does prevent execution of:
```
init of catalogs
prune catalogs
```
or
```
prune catalogs
init of catalogs
```
lock on it's own doesn't enforce order. Do I miss something?

Reply:
[This is the PR for it:](https://github.com/trinodb/trino/pull/19683) was able to replicate in unit test. The general order was that while prune was collecting the catalogs for a task, it was possible for a task to be assigned catalogs, and load them, only for the pruning to run right after before the query executed.

Reply:
> https://github.com/trinodb/trino/pull/19683 was able to replicate in unit test.

How does that PR relates to change here?

Reply:
That PR introduced the lock. This PR just move the location of the lock